request_handlers.cc

```diff
@@ -181,15 +185,217 @@ RequestHandler::Status ReverseProxyRequestHandler::handleRequest(const Request&
+    redirect.AddHeader("Connection", "close") ; 
+
+    std::string redirect_request = redirect.ToString(); 
+    std::cout << "=== REDIRECT REQUEST ===\n" << redirect_request << std::endl;
+  
+    // Send the redirect request: 
+    boost::asio::write(socket,
+                       boost::asio::buffer(redirect_request.c_str(),
+                                           redirect_request.size()));
+    // Receive redirect response
+    std::string raw_redirect_response; 
+    if (!ReverseProxyRequestHandler::obtainResponse(socket, raw_redirect_response)) {
+      std::cout << "Response from redirect failed to be read.\n";
+      return RequestHandler::NOT_FOUND;
+    }
+    
+    // Parse raw redirect response 
+    auto parsed_redirect_response = Response::Parse(raw_redirect_response);
+    if (parsed_redirect_response == nullptr) {
+      std::cout << "An error occured parsing the response.\n";
+      return RequestHandler::NOT_FOUND;
+    }
+
+    if (parsed_redirect_response->statusCode() == "302" || parsed_redirect_response->statusCode() == "301") {
+      std::cout << "Attempting to redirect request more than once.\n";
+      return RequestHandler::NOT_FOUND;
+    }
+
+    // Request successful, write response to server
+    std::string parsed_redirect_resp_str = parsed_redirect_response->ToString();
+    std::cout << "=== REDIRECT RESPONSE FROM REMOTE ===\n" << parsed_redirect_resp_str << std::endl;
+
+    // Some pointer magic to convert a unique_ptr to a raw pointer 
+    *(response) = *(parsed_redirect_response.get());
+
+    return RequestHandler::OK;
   
+  }
+
+  // Request successful, write response to server
+  // Some pointer magic to convert a unique_ptr to a raw pointer 
+  *(response) = *(parsed_response.get());
+
   return RequestHandler::OK;
 }
 
+// Takes in an uri of the following format: /proxy_path/foo/bar
+// and obtains the path following /proxy_path => /foo/bar
+// 
+// If nothing follows the uri like so: /proxy_path
+// The function will return an empty string
+
+std::string ReverseProxyRequestHandler::obtainPath(std::string uri) {
+  // Assumption: non-root paths for ReverseProxyRequestHandler will be
+  // prefixed with the word "proxy" in its path
+  // IE /reverse_proxy_path, /proxy, ect
+  
+  // Handle case where files are served from "/"
+  // IE /static/index.html, return the uri 
+  boost::iterator_range<string::iterator> find_proxy = boost::find_nth(uri, "proxy", 0); 
+  if (!find_proxy) {
+    return uri; 
+  }
+  std::string default_path = "";
+  boost::iterator_range<string::iterator> r = boost::find_nth(uri, "/", 1);
+  
+  if (r) {
+    std::string path(r.begin(), uri.end());
+    return path;
+  }
+  return default_path;
+}
+
+// If a remote proxy host is in the form: http:://www.foo.com/ return
+// www.foo.com instead. 
+std::string ReverseProxyRequestHandler::sanitizeProxyHost(std::string& remote_proxy_host) {
+
+  // Remove trailing forward slash 
+  if (remote_proxy_host.back() == '/') {
```

Comment:
We've decided that while this will be more robust, handling more complex remote proxy hosts is out of scope for this project. Thank you for pointing this out though!